src/main/java/org/jenkinsci/plugins/workflow/cps/CpsGroovyShell.java

```diff
@@ -28,9 +35,51 @@
     /**
      * Use {@link CpsGroovyShellFactory} to instantiate it.
      */
+    @SuppressFBWarnings(value="DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED", justification="irrelevant")
     CpsGroovyShell(ClassLoader parent, @CheckForNull CpsFlowExecution execution, CompilerConfiguration cc) {
         super(parent,new Binding(),cc);
         this.execution = execution;
+        try {
+            // Apparently there is no supported way to initialize a GroovyShell with a specified GroovyClassLoader.
+            Field loaderF = GroovyShell.class.getDeclaredField("loader");
+            loaderF.setAccessible(true);
+            loaderF.set(this, new CleanGroovyClassLoader(parent, cc));
+        } catch (Exception x) {
+            LOGGER.log(Level.WARNING, "failed to install CleanGroovyClassLoader", x);
+        }
+    }
+
+    /**
+     * Disables the weird and unreliable {@link groovy.lang.GroovyClassLoader.InnerLoader}.
+     * This is apparently only necessary when you are using class recompilation, which we are not.
+     * We want the {@linkplain Class#getClassLoader defining loader} of {@code *.groovy} to be this one.
+     * Otherwise the defining loader will be an {@code InnerLoader}, and not necessarily the same instance from load to load.
+     * @see GroovyClassLoader#getTimeStamp
+     */
+    private static final class CleanGroovyClassLoader extends GroovyClassLoader {
```

Comment:
Ah, okay so the big win here is that we're reusing the loader itself and not creating multiple ones for compilations of different SourceUnits.  

I can't help but feel like this opens some risk that one of the implementations relies on some separation of InnerLoaders (for example using the same class name for different SourceUnits and relying on different loaders  to distinguish them).  Have you looked closely at the implications of this?

But on the other hand: this will probably reduce memory footprint, implement a more "correct" approach and may actually end (or simplify debugging) PermGen memory leaks.